File: java/src/Solution0290.java
import java.util.HashMap;
import java.util.Map;

public class Solution0290 {
    public boolean wordPattern(String pattern, String s) {
        String[] ss = s.split(" ");
        if (pattern.length() != ss.length) {
            return false;
        }
        Map<Character, String> map = new HashMap<>();
        Map<String, Character> map2 = new HashMap<>();
        for (int i = 0; i < pattern.length(); i++) {
            char tc = pattern.charAt(i);
            String ts = ss[i];
            if (map.containsKey(tc)) {
                if (!map.get(tc).equals(ts)) return false;
            } else {
                //tc 和ts对应，但是存在一个ts，不和tc对应，错误
                if (map2.containsKey(ts) && map2.get(ts) != tc) {
                    return false;
                }
                map.put(tc, ts);
                map2.put(ts, tc);
            }
        }
        return true;
    }
}

File: goR3/solution0290.go
package main

import "strings"

func WordPattern(pattern string, s string) bool {
	m1 := make(map[byte]string)
	m2 := make(map[string]byte)

	s1 := strings.Split(s, " ")

	if len(pattern) != len(s1) {
		return false
	}

	for i := 0; i < len(s1); i++ {
		_, ok := m1[pattern[i]]
		if ok {
			if m1[pattern[i]] != s1[i] {
				return false
			}
		} else {
			m1[pattern[i]] = s1[i]
		}

		_, ok2 := m2[s1[i]]
		if ok2 {
			if m2[s1[i]] != pattern[i] {
				return false
			}
		} else {
			m2[s1[i]] = pattern[i]
		}
	}

	return true
}